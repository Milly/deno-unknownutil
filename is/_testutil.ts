// It seems 'IsExact' in deno_std is false positive so use `Equal` in type-challenges
// https://github.com/type-challenges/type-challenges/blob/e77262dba62e9254451f661cb4fe5517ffd1d933/utils/index.d.ts#L7-L9
export type Equal<X, Y> = (<T>() => T extends X ? 1 : 2) extends
  (<T>() => T extends Y ? 1 : 2) ? true : false;

export function stringify(x: unknown): string {
  if (x instanceof Date) return `Date(${x.valueOf()})`;
  if (x instanceof Promise) return "Promise";
  if (typeof x === "function") return x.toString();
  if (typeof x === "bigint") return `${x}n`;
  if (typeof x === "symbol") return x.toString();
  return JSON.stringify(x);
}
