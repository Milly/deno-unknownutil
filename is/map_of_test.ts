import { assertEquals } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import { assertType } from "@std/testing/types";
import { type Equal, testWithExamples } from "../_testutil.ts";
import { is } from "./mod.ts";
import { isMapOf } from "./map_of.ts";

Deno.test("isMapOf<T>", async (t) => {
  await t.step("returns properly named function", async (t) => {
    await assertSnapshot(t, isMapOf(is.Number).name);
    await assertSnapshot(t, isMapOf((_x): _x is string => false).name);
  });
  await t.step("returns proper type predicate", () => {
    const a: unknown = new Map([["a", 0]]);
    if (isMapOf(is.Number)(a)) {
      assertType<Equal<typeof a, Map<unknown, number>>>(true);
    }
  });
  await t.step("returns true on T map", () => {
    assertEquals(isMapOf(is.Number)(new Map([["a", 0]])), true);
    assertEquals(isMapOf(is.String)(new Map([["a", "a"]])), true);
    assertEquals(isMapOf(is.Boolean)(new Map([["a", true]])), true);
  });
  await t.step("returns false on non T map", () => {
    assertEquals(isMapOf(is.String)(new Map([["a", 0]])), false);
    assertEquals(isMapOf(is.Number)(new Map([["a", "a"]])), false);
    assertEquals(isMapOf(is.String)(new Map([["a", true]])), false);
  });
  await testWithExamples(t, isMapOf((_: unknown): _ is unknown => true), {
    excludeExamples: ["map"],
  });
});

Deno.test("isMapOf<T, K>", async (t) => {
  await t.step("returns properly named function", async (t) => {
    await assertSnapshot(t, isMapOf(is.Number, is.String).name);
    await assertSnapshot(
      t,
      isMapOf((_x): _x is string => false, is.String).name,
    );
  });
  await t.step("returns proper type predicate", () => {
    const a: unknown = new Map([["a", 0]]);
    if (isMapOf(is.Number, is.String)(a)) {
      assertType<Equal<typeof a, Map<string, number>>>(true);
    }
  });
  await t.step("returns true on T map", () => {
    assertEquals(isMapOf(is.Number, is.String)(new Map([["a", 0]])), true);
    assertEquals(isMapOf(is.String, is.String)(new Map([["a", "a"]])), true);
    assertEquals(isMapOf(is.Boolean, is.String)(new Map([["a", true]])), true);
  });
  await t.step("returns false on non T map", () => {
    assertEquals(isMapOf(is.String, is.String)(new Map([["a", 0]])), false);
    assertEquals(isMapOf(is.Number, is.String)(new Map([["a", "a"]])), false);
    assertEquals(isMapOf(is.String, is.String)(new Map([["a", true]])), false);
  });
  await t.step("returns false on non K map", () => {
    assertEquals(isMapOf(is.Number, is.Number)(new Map([["a", 0]])), false);
    assertEquals(isMapOf(is.String, is.Number)(new Map([["a", "a"]])), false);
    assertEquals(isMapOf(is.Boolean, is.Number)(new Map([["a", true]])), false);
  });
  await testWithExamples(t, isMapOf((_: unknown): _ is unknown => true), {
    excludeExamples: ["map"],
  });
});
