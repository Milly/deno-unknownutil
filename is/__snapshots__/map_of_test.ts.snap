export const snapshot = {};

snapshot[`isMapOf<T> > returns properly named function 1`] = `"isMapOf(isNumber, undefined)"`;

snapshot[`isMapOf<T> > returns properly named function 2`] = `"isMapOf((anonymous), undefined)"`;

snapshot[`isMapOf<T, K> > returns properly named function 1`] = `"isMapOf(isNumber, isString)"`;

snapshot[`isMapOf<T, K> > returns properly named function 2`] = `"isMapOf((anonymous), isString)"`;
